assets/man/borgreport.1
'\" t
.\"     Title: borgreport
.\"    Author: [see the "AUTHOR(S)" section]
.\" Generator: Asciidoctor 2.0.23
.\"      Date: 2025-06-08
.\"    Manual: borgreport User Manual
.\"    Source: borgreport 0.3.0
.\"  Language: English
.\"
.TH "BORGREPORT" "1" "2025-06-08" "borgreport 0.3.0" "borgreport User Manual"
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.ss \n[.ss] 0
.nh
.ad l
.de URL
\fI\\$2\fP <\\$1>\\$3
..
.als MTO URL
.if \n[.g] \{\
.  mso www.tmac
.  am URL
.    ad l
.  .
.  am MTO
.    ad l
.  .
.  LINKSTYLE blue R < >
.\}
.SH "NAME"
borgreport \- Summarize the status of multiple BorgBackup repositories in one report and export metrics

.SH "SYNOPSIS"
.sp
\fBborgreport\fP
[\-\-env\-dir <PATH>]
[\-\-env\-inherit <REPOSITORY>]
[\-\-text\-to <FILE>]
[\-\-html\-to <FILE>]
[\-\-metrics\-to <FILE>]
[\-\-mail\-to <ADDR>]
[\-\-mail\-from <ADDR>]
[\-\-no\-progress]
[\-\-glob\-archives <GLOB>]
[\-\-check]
[\-\-check\-options <OPTS>]
[\-\-compact]
[\-\-compact\-options <OPTS>]
[\-\-borg\-binary]
[\-\-max\-age\-hours]
[\-h|\-\-help]
[\-V|\-\-version]
.SH "DESCRIPTION"
.sp
A wrapper for BorgBackup to query the latest backup archives and perform health checks on repositories. It summarize the status of BorgBackup repositories with statistics, warnings and error messages. You can save the report as file or send it per mail and export OpenMetrics (Prometheus Metrics) for the last archives.
.sp
borgreport takes a directory with *.env files as input. Each file must contain environment variables as understood by BorgBackup to access a repository. The filename will be shown as name of the repository in the report.
.sp
borgreport can inherit BORG_* env vars for a single repository. This allows to run borgreport after BorgBackup while reusing the environment.
.SH "OPTIONS"
.sp
\-\-env\-dir=<PATH>
.RS 4
Directory  to look for \fI*.env\fP files containing BORG_* environment variables for a repository. Each file name represents a repository name in the report.
.RE
.sp
\-\-env\-inherit=<REPOSITORY>
.RS 4
Inherit BORG_* environment variables for a single repository. This allows to run \f(CRborgreport\fP after \f(CRborg\fP while reusing the environment. <REPOSITORY> will be shown as the name in the report.
.RE
.sp
\-\-text\-to=<FILE>
.RS 4
Write the text report to <FILE> instead of stdout.
.RE
.sp
\-\-html\-to=<FILE>
.RS 4
Write the HTML report to <FILE>.
.RE
.sp
\-\-metrics\-to=<FILE>
.RS 4
Write metrics to <FILE>.
.RE
.sp
\-\-mail\-to=<ADDR>
.RS 4
Send the report to <ADDR> using a \f(CRsendmail\fP compatible mail transfer agent.
.RE
.sp
\-\-mail\-from=<ADDR>
.RS 4
The mail sender <ADDR>. By default this is the current user@host
.RE
.sp
\-\-no\-progress
.RS 4
Suppress all status updates during processing. By default this is auto\-detected.
.RE
.sp
\-h, \-\-help
.RS 4
Print help (see a summary with \f(CR\-h\fP)
.RE
.sp
\-V, \-\-version
.RS 4
Print version (see a summary with \f(CR\-V\fP)
.RE
.SS "Repository options"
.sp
The following options change the behavior for all repositories. To configure a single repository set an environment variable in the respective \fI*.env\fP file.
.sp
\-\-glob\-archives=<GLOB>
.RS 4
A list of space separated archive globs e.g. "etc\-* srv\-*" for archive names starting with etc\- or srv\-.  (Default: "")
.RE
.sp
\-\-check[=<true|false>]
.RS 4
Enables the execution of \f(CRborg check\fP. (Default: false)
.RE
.sp
\-\-check\-options=<OPTS>
.RS 4
A list of space separated raw borg options supplied to the \f(CRborg check\fP command.
.RE
.sp
\-\-compact[=<true|false>]
.RS 4
Enables the execution of \f(CRborg compact\fP. (Default: false)
.RE
.sp
\-\-compact\-options=<OPTS>
.RS 4
A list of space separated raw borg options supplied to the \f(CRborg compact\fP command.
.RE
.sp
\-\-borg\-binary=<FILE>
.RS 4
Path to a local \f(CRborg\fP binary. (Default: borg)
.RE
.sp
\-\-max\-age\-hours=<HOURS>
.RS 4
Threshold to warn, when the last backup is older than <HOURS>. (Default: 24)
.RE
.SH "ENVIRONMENT"
.sp
Environment variables are overwritten by the respective command line option.
.sp
BORGREPORT_ENV_DIR <PATH>
.RS 4
Directory to look for \fI*.env\fP files containing BORG_* environment variables for a repository. Each file name represents a repository name in the report.
.RE
.sp
BORGREPORT_ENV_INHERIT <REPOSITORY>
.RS 4
Inherit BORG_* environment variables for a single <REPOSITORY>. This allows to run \f(CRborgreport\fP after \f(CRborg\fP while reusing the environment.
.RE
.sp
BORGREPORT_MAIL_TO <ADDR>
.RS 4
Send the report to <ADDR> using a \f(CRsendmail\fP compatible mail transfer agent.
.RE
.sp
BORGREPORT_MAIL_FROM <ADDR>
.RS 4
The mail sender <ADDR>. By default this is the \f(CRuser@host\fP.
.RE
.sp
BORGREPORT_NO_PROGRESS <ADDR>
.RS 4
Suppress all status updates during processing. By default this is auto\-detected.
.RE
.sp
BORGREPORT_TEXT_TO <FILE>
.RS 4
Write the text report to <FILE> instead of stdout.
.RE
.sp
BORGREPORT_HTML_TO <FORMAT>
.RS 4
Write the HTML report to <FILE>.
.RE
.sp
BORGREPORT_METRICS_TO <FILE>
.RS 4
Write metrics to <FILE>.
.RE
.SS "Repository Environment"
.sp
Set the following variables at repository level as part of the \fI*.env\fP file. Setting them globally will alter the default behavior for all repositories.
.sp
BORGREPORT_GLOB_ARCHIVES <GLOB>
.RS 4
A list of space separated archive globs e.g. "etc\-* srv\-*" for archive names starting with etc\- or srv\-. (Default: "")
.RE
.sp
BORGREPORT_CHECK <true|false>
.RS 4
Enables the execution of ‘borg check‘. (Default: false)
.RE
.sp
BORGREPORT_CHECK_OPTIONS <OPTS>
.RS 4
A list of space separated raw borg options supplied to the \f(CRborg check\fP command.q
.RE
.sp
BORGREPORT_COMPACT <true|false>
.RS 4
Enables the execution of \f(CRborg compact\fP. (Default: false)
.RE
.sp
BORGREPORT_COMPACT_OPTIONS <OPTS>
.RS 4
A list of space separated raw borg options supplied to the \f(CRborg compact\fP command.
.RE
.sp
BORGREPORT_BORG_BINARY <FILE>
.RS 4
Path to a local \f(CRborg\fP binary. (Default: borg)
.RE
.sp
BORGREPORT_MAX_AGE_HOURS <HOURS>
.RS 4
Threshold to warn, when the last backup is older than <HOURS>. (Default: 24)
.RE
.SH "EXAMPLES"
.SS "Use a directory with *.env files to describe repositories"
.sp
.if n .RS 4
.nf
.fam C
# Create an env file with the BORG_* variables for each repo. The filename is shown as the repo name in the report.
$ mkdir repos
$ cat repos/somerepo.env
BORG_REPO=/mnt/borg/repos/somerepo
BORG_PASSPHRASE=Secure

# Print the report to stdout and run `borg check` against the repos
borgreport \-\-env\-dir repos \-\-check

# Send the report via `sendmail` to admin@host.invalid
borgreport \-\-env\-dir repos \-\-mail\-to admin@example.com

# Write the metrics to file borg.metrics and print a text report to stdout
borgreport \-\-env\-dir repos \-\-metrics\-to borg.metrics \-\-text\-to=\-
.fam
.fi
.if n .RE
.SS "Use BORG_* variables for a single repository from the environment"
.sp
.if n .RS 4
.nf
.fam C
export BORG_REPO=/mnt/borg/repos/somerepo
export BORG_PASSPHRASE=Secure

# Run borg as normal
borg create borg create \*(Aq::{utcnow}\*(Aq /data

# Export the metrics for the recent archive to file borg.metric (and parse the repo name from BORG_REPO).
borgreport \-\-metrics\-to borg.metrics

# Export the metrics for the recent archives starting with etc\- or srv\- to file borg.metrics (and name the repo "somerepo").
borgreport \-\-env\-inherit somerepo \-\-glob\-archives \*(Aqetc\-* srv\-*\*(Aq \-\-metrics\-to borg.metrics
.fam
.fi
.if n .RE
.SH "AUTHOR"
.sp
Written by Philipp Micheel \c
.MTO "bbx0+borgreport\(atbitdevs.de" "" ""
.SH "REPORTING BUGS"
.sp
Report bugs to \c
.URL "https://github.com/bbx0/borgreport/issues" "" ""
.SH "COPYRIGHT"
.sp
Copyright \(co 2024 Philipp Micheel. License GPLv3+: GNU GPL version 3 or later \c
.URL "https://gnu.org/licenses/gpl.html" "" "."
.br
There is ABSOLUTELY NO WARRANTY, to the extent permitted by law.
.SH "SEE ALSO"
.sp
borg(1)

CHANGELOG.md
<!-- SPDX-FileCopyrightText: 2024 Philipp Micheel <bbx0+borgreport@bitdevs.de> -->
<!-- SPDX-License-Identifier: CC-BY-SA-4.0 -->
<!-- markdownlint-configure-file {"MD024": { "siblings_only": true } } -->
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased](https://github.com/bbx0/borgreport/compare/v0.3.0...HEAD) - 2025-06-08

### Added

- Option `--compact` to perform `borg compact` on a repository.

### Changed

- Use `asciidoc` instead of `help2man` to create the manpage
- Refactored integration tests from `tests.sh` into `cargo test`
- Refactored for consistent use of `jiff::Zoned` and `jiff::SignedDuration`
- Adopted changes from REUSE specification 3.3
- Updated to rust edition 2024 which raises MSRV to 1.85
- Updated dependencies (chore)

## [0.3.0](https://github.com/bbx0/borgreport/compare/v0.2.0...v0.3.0) - 2024-11-12

### Added

- [OpenMetrics](https://github.com/prometheus/OpenMetrics/blob/main/specification/OpenMetrics.md) Text Exporter (Prometheus Metrics)
- Option `--metrics-to` to save metrics in `application/openmetrics-text` format
- Option `--env-inherit` to inherit BORG_* variables for a single repository from the active environment. This allows to run `borgreport` after `borg` while reusing the environment.
- Options `--text-to` and `--html-to` to save the report in `text/plain` or  `text/html` format
- Option `--check-options` to supply raw `borg check` options, when `--check` is enabled.

### Changed

- Replaced the tmpfiles.d configuration with systemd unit directives to manage the config and state directories.
- When called without `--env-inherit` and `--env-dir` but a `BORG_REPO` environment variable is provided, the repository name is set to the final component of the `BORG_REPO` path. This is a convenience feature. It is recommended to use `--env-inherit` to name the repository explicitly or to use a `--env-dir`.

### Removed

- The options `--file-to` and `--file-format` are removed and replaced by individual flags: `--text-to`, `--html-to`.

## [0.2.0](https://github.com/bbx0/borgreport/compare/v0.1.0...v0.2.0) - 2024-10-13

### Added

- Support to format the report in HTML
- Option `--file-format` to choose the output file format

### Changed

- Use `--bypass-lock` for the `borg info` command
- Send Emails as multipart with `text/plain` and `text/html`

## [0.1.0](https://github.com/bbx0/borgreport/releases/tag/v0.1.0) - 2024-10-02

### Added

- Summarize status of BorgBackup repositories with statistics, warnings and error messages.
- Perform simple sanity checks
  - Warn about empty backup sources or repositories
  - Warn if the age of the last backup exceeds a threshold (24 hours by default)
- Send reports via Email
- Allow to run `borg check` as part of the report.
- Integrate with systemd as service and timer


assets/shell_completions/borgreport.bash
_borgreport() {
    local i cur prev opts cmd
    COMPREPLY=()
    cur="${COMP_WORDS[COMP_CWORD]}"
    prev="${COMP_WORDS[COMP_CWORD-1]}"
    cmd=""
    opts=""

    for i in ${COMP_WORDS[@]}
    do
        case "${cmd},${i}" in
            ",$1")
                cmd="borgreport"
                ;;
            *)
                ;;
        esac
    done

    case "${cmd}" in
        borgreport)
            opts="-h -V --env-dir --env-inherit --text-to --html-to --metrics-to --mail-to --mail-from --no-progress --glob-archives --check --check-options --borg-binary --max-age-hours --help-man --help --version"
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 1 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                --env-dir)
                    COMPREPLY=()
                    if [[ "${BASH_VERSINFO[0]}" -ge 4 ]]; then
                        compopt -o plusdirs
                    fi
                    return 0
                    ;;
                --env-inherit)
                    COMPREPLY=("${cur}")
                    if [[ "${BASH_VERSINFO[0]}" -ge 4 ]]; then
                        compopt -o nospace
                    fi
                    return 0
                    ;;
                --text-to)
                    local oldifs
                    if [ -n "${IFS+x}" ]; then
                        oldifs="$IFS"
                    fi
                    IFS=$'\n'
                    COMPREPLY=($(compgen -f "${cur}"))
                    if [ -n "${oldifs+x}" ]; then
                        IFS="$oldifs"
                    fi
                    if [[ "${BASH_VERSINFO[0]}" -ge 4 ]]; then
                        compopt -o filenames
                    fi
                    return 0
                    ;;
                --html-to)
                    local oldifs
                    if [ -n "${IFS+x}" ]; then
                        oldifs="$IFS"
                    fi
                    IFS=$'\n'
                    COMPREPLY=($(compgen -f "${cur}"))
                    if [ -n "${oldifs+x}" ]; then
                        IFS="$oldifs"
                    fi
                    if [[ "${BASH_VERSINFO[0]}" -ge 4 ]]; then
                        compopt -o filenames
                    fi
                    return 0
                    ;;
                --metrics-to)
                    local oldifs
                    if [ -n "${IFS+x}" ]; then
                        oldifs="$IFS"
                    fi
                    IFS=$'\n'
                    COMPREPLY=($(compgen -f "${cur}"))
                    if [ -n "${oldifs+x}" ]; then
                        IFS="$oldifs"
                    fi
                    if [[ "${BASH_VERSINFO[0]}" -ge 4 ]]; then
                        compopt -o filenames
                    fi
                    return 0
                    ;;
                --mail-to)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --mail-from)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --glob-archives)
                    COMPREPLY=("${cur}")
                    if [[ "${BASH_VERSINFO[0]}" -ge 4 ]]; then
                        compopt -o nospace
                    fi
                    return 0
                    ;;
                --check)
                    COMPREPLY=($(compgen -W "true false" -- "${cur}"))
                    if [[ "${BASH_VERSINFO[0]}" -ge 4 ]]; then
                        compopt -o nospace
                    fi
                    return 0
                    ;;
                --check-options)
                    COMPREPLY=("${cur}")
                    if [[ "${BASH_VERSINFO[0]}" -ge 4 ]]; then
                        compopt -o nospace
                    fi
                    return 0
                    ;;
                --borg-binary)
                    local oldifs
                    if [ -n "${IFS+x}" ]; then
                        oldifs="$IFS"
                    fi
                    IFS=$'\n'
                    COMPREPLY=($(compgen -f "${cur}"))
                    if [ -n "${oldifs+x}" ]; then
                        IFS="$oldifs"
                    fi
                    if [[ "${BASH_VERSINFO[0]}" -ge 4 ]]; then
                        compopt -o filenames
                    fi
                    return 0
                    ;;
                --max-age-hours)
                    COMPREPLY=("${cur}")
                    if [[ "${BASH_VERSINFO[0]}" -ge 4 ]]; then
                        compopt -o nospace
                    fi
                    return 0
                    ;;
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
    esac
}

if [[ "${BASH_VERSINFO[0]}" -eq 4 && "${BASH_VERSINFO[1]}" -ge 4 || "${BASH_VERSINFO[0]}" -gt 4 ]]; then
    complete -F _borgreport -o nosort -o bashdefault -o default borgreport
else
    complete -F _borgreport -o bashdefault -o default borgreport
fi
